File: examples/python/rest/greeks.py
"""Copyright 2026 Volstrata.com

The `greeks` and `vol` domains, and what a plan floor looks like from a client.
Docs: https://volstrata.com/docs/api-catalog
      https://volstrata.com/docs/api-auth

    python rest/greeks.py

Four capabilities, two open and two gated on purpose:

    greeks.gamma   GET /api/v1/greeks/gamma    per-strike gamma exposure
    vol.iv         GET /api/v1/vol/iv          at-the-money implied vol
    greeks.charm   GET /api/v1/greeks/charm    above the lowest floor
    vol.surface    GET /api/v1/vol/surface     above the lowest floor

The open two also carry the two shared blocks worth understanding once, because
they appear across the surface:

    freshness  {as_of, age_seconds, live, stale, stale_after_seconds}
    delivery   {mode: "live"|"delayed", delay_seconds, reason}

`delivery.reason` is "plan_tier" when the delay is a function of your plan and
"source_delay" when it is not, so a client can tell "upgrade to fix this" from
"nobody has it faster". `delay_seconds` is 0 when the body really is realtime and
null when it is delayed by an amount the server cannot state -- never coerce null
to 0, and branch on None explicitly.

Four sequential requests. Two of them are expected to be refused.
"""

import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from volstrata_helpers import (  # noqa: E402  (import after the path fix, on purpose)
    attempt,
    describe_freshness,
    get,
    heading,
    key_banner,
    row,
    run_examples,
    show,
)

TICKER = "SPX"


def gamma_by_strike() -> None:
    """greeks.gamma returns the per-strike series plus the scalars around it.

    Shape:

        {"greek": "gamma",
         "strikes": [<strike>, <strike>, ...],   the strike axis
         "bars": [[<value>, <strike>], ...],     value first, strike second
         "window": [<low>, <high>],              the strike window covered
         "net_bn": <number>,                     magnitude in billions
         "peak_pos_strike": ..., "peak_neg_strike": ...,
         "regime": "...", "regime_basis": "...",
         "spot": ..., "ts": ..., "updated": "...",
         "freshness": {...}, "delivery": {...}}

    `strikes` and `bars` are two views of the same axis: use `bars` when you want
    a value and its strike together, `strikes` when you only need the axis. Mind
    the order inside a bar -- the value comes first and the strike second, which
    is the opposite of what most people assume on sight.
    """
    heading("greeks.gamma -- per-strike gamma exposure")
    data = get("/greeks/gamma", ticker=TICKER)
    row("greek", data.get("greek"))
    row("spot", data.get("spot"))
    row("net_bn", data.get("net_bn"))
    row("peak_pos_strike", data.get("peak_pos_strike"))
    row("peak_neg_strike", data.get("peak_neg_strike"))
    row("regime", data.get("regime"))
    row("regime_basis", data.get("regime_basis"))
    row("window", data.get("window"))
    bars = data.get("bars") or []
    row("bars", "{0} [value, strike] pairs".format(len(bars)))
    if bars:
        show(bars[:3])
    strikes = data.get("strikes") or []
    if strikes:
        row("strike axis", "{0} points, {1} .. {2}".format(
            len(strikes), strikes[0], strikes[-1]))
    row("freshness", describe_freshness(data))
    print("  A `delivery.mode` of 'delayed' with reason 'plan_tier' means the")
    print("  same call on a higher plan returns a fresher frame, not more fields.")


def implied_vol() -> None:
    """vol.iv is the compact implied-volatility read.

        {"atm_iv_today": <fraction>, "atm_iv_weekly": <fraction>,
         "iv_axis": [<low>, <high>],
         "surface": {"has_weekly": true, "smile0_points": <n>, "smile1_points": <n>},
         "spot": ..., "ts": ..., "updated": "...",
         "freshness": {...}, "delivery": {...}}

    Values are decimal fractions, not percentages: 0.14 is 14%. `surface` is a
    description of what the fuller surface capability would return for this
    moment -- point counts and whether a weekly smile exists -- not the surface
    itself.
    """
    heading("vol.iv -- at-the-money implied volatility")
    data = get("/vol/iv", ticker=TICKER)
    row("atm_iv_today", data.get("atm_iv_today"))
    row("atm_iv_weekly", data.get("atm_iv_weekly"))
    row("iv_axis", data.get("iv_axis"))
    row("surface", data.get("surface"))
    row("freshness", describe_freshness(data))


def gated_charm() -> None:
    """greeks.charm sits above the lowest floor. Catch, report, continue.

    The `greeks` domain publishes eleven capabilities and gamma is the one at
    the Free floor; the higher-order greeks sit above it. Discover which is
    which at runtime rather than hard-coding a list:

        GET /api/v1/meta/access   -> {"unlocked": [...], "locked": [...]}
        each locked row carries min_tier / min_tier_name

    discover_capabilities.py prints exactly that.
    """
    heading("greeks.charm -- gated")
    data = attempt("greeks.charm", get, "/greeks/charm", ticker=TICKER)
    if data is None:
        print("  Expected without the plan. The refusal names what would clear it.")
        return
    row("keys", ", ".join(sorted(data)[:12]))
    row("freshness", describe_freshness(data))


def gated_surface() -> None:
    """vol.surface is the full surface, and also above the lowest floor.

    Note the parameters it declares (ticker, window, expiry, format): a gated
    capability is fully described in the public spec even when your plan cannot
    call it, so you can build against it before you buy it.
    """
    heading("vol.surface -- gated")
    data = attempt("vol.surface", get, "/vol/surface", ticker=TICKER)
    if data is None:
        print("  Its parameters are still public: GET /api/openapi.json documents")
        print("  every operation regardless of your plan.")
        return
    row("keys", ", ".join(sorted(data)[:12]))


def main() -> int:
    print("VolStrata: greeks and vol -- https://volstrata.com")
    key_banner()
    exit_code = run_examples((
        ("gamma by strike", gamma_by_strike),
        ("implied vol", implied_vol),
        ("charm (gated)", gated_charm),
        ("surface (gated)", gated_surface),
    ))
    heading("Next")
    print("  rest/gex_levels.py   the levels these greeks sit behind")
    print("  auth_and_errors.py   the refusal envelope in full")
    print("  Plans and keys: https://volstrata.com/api-keys")
    return exit_code


if __name__ == "__main__":
    sys.exit(main())

File: examples/javascript/rest/greeks.mjs
/**
 * Copyright 2026 Volstrata.com
 *
 * The `greeks.*` and `vol.*` domains: one Free capability and two that sit above
 * the Free floor, so this file is also the shortest demonstration of how a
 * mixed-tier script should behave.
 * Catalog: https://volstrata.com/docs/api-catalog
 *
 * Run:  node rest/greeks.mjs [TICKER]
 *       npm run greeks
 *
 * Three requests:
 *
 *   greeks.gamma    Free   — always runs
 *   greeks.charm    Pro    — refuses without the plan, and says so
 *   vol.surface     Pro    — same
 *
 * Two of three rows refusing is the expected, successful outcome of running this
 * with no key. The exit code is 0 either way: the script's job is to report what
 * the API said, and "you need the Pro plan for this" is something the API said.
 */

import {
  apiFetch,
  apiGet,
  banner,
  callCapability,
  hasApiKey,
  heading,
  keysOf,
  report,
  show,
  showRateLimit,
  summarize,
} from "../lib/volstrata.mjs";

const ticker = (process.argv[2] ?? "SPX").toUpperCase();

banner(`greeks.* and vol.* — ${ticker}`);
show("ticker", `${ticker} (pass another as the first argument)`);

const results = [];

// ---------------------------------------------------------------------------
// greeks.gamma — Free
// ---------------------------------------------------------------------------
//
// The one row here that runs for everybody. Field names are not restated in this
// comment on purpose: read them off the response, or look the metric up in the
// glossary. `docs.metric` (see rest/research_and_docs.mjs) returns the
// definition of any metric name over the API itself.

results.push(
  await callCapability("greeks.gamma", async () => {
    const { res, data } = await apiFetch("/api/v1/greeks/gamma", { ticker });

    show("keys", keysOf(data));

    // Most of these payloads are either a ladder (one row per strike) or a small
    // map of scalars. Find whichever this is instead of assuming.
    const [rowsKey, rows] = Object.entries(data).find(([, v]) => Array.isArray(v)) ?? [];
    if (rows) {
      show("rows key", rowsKey);
      show("rows", rows.length);
      show("first row", summarize(rows[0], { depth: 1, maxKeys: 8 }));
    } else {
      show("payload", summarize(data, { depth: 2, maxKeys: 12 }));
    }

    showRateLimit(res);
    return data;
  }, { note: "Free — the higher-order greeks below are not" }),
);

// ---------------------------------------------------------------------------
// greeks.charm — Pro
// ---------------------------------------------------------------------------
//
// `window` and `at` are the two parameters worth knowing across this domain:
// `window` selects the session ("today" is the default) and `at` selects the
// point in it ("latest" by default). Both are passed explicitly here because
// being explicit is the habit worth having in your own code.

results.push(
  await callCapability("greeks.charm", async () => {
    const data = await apiGet("/api/v1/greeks/charm", { ticker, window: "today", at: "latest" });
    show("keys", keysOf(data));
    show("payload", summarize(data, { depth: 2, maxKeys: 10 }));
    return data;
  }, { note: "Pro plan" }),
);

// ---------------------------------------------------------------------------
// vol.surface — Pro
// ---------------------------------------------------------------------------
//
// `expiry` narrows the surface to a single expiry when you pass one; leaving it
// off returns what the capability considers the default view. `lib/volstrata.mjs`
// drops parameters whose value is undefined, so an unset option simply does not
// appear on the query string.

results.push(
  await callCapability("vol.surface", async () => {
    const data = await apiGet("/api/v1/vol/surface", { ticker, window: "today", expiry: undefined });
    show("keys", keysOf(data));
    show("payload", summarize(data, { depth: 2, maxKeys: 10 }));
    return data;
  }, { note: "Pro plan" }),
);

report(results);

heading("How a refusal should read");
if (hasApiKey) {
  console.log("  A key is set. Rows that still refused above are a plan floor, not a");
  console.log("  credential problem — the problem document names the plan required.");
} else {
  console.log("  With no key, the two Pro rows above answered 401 auth_required.");
  console.log("  With a key below the floor they answer with required_plan_name set");
  console.log("  instead, which is the field to show a human.");
}

heading("Related");
console.log("  Capability catalog   https://volstrata.com/docs/api-catalog");
console.log("  Authentication       https://volstrata.com/docs/api-auth");
console.log("  Next domain          node rest/valuation.mjs");
